runtime-modules/project-token/src/lib.rs

```diff
@@ -539,24 +773,43 @@ impl<T: Trait> Module<T> {
 
         let cumulative_bloat_bond = Self::compute_bloat_bond(validated_transfers);
         if !cumulative_bloat_bond.is_zero() {
-            let treasury_account_id = Self::bloat_bond_treasury_account_id();
-            let _ = T::ReserveCurrency::transfer(
-                src,
-                &treasury_account_id,
-                cumulative_bloat_bond,
-                ExistenceRequirement::KeepAlive,
-            );
+            Self::deposit_to_treasury(src, cumulative_bloat_bond);
         }
 
         AccountInfoByTokenAndAccount::<T>::mutate(token_id, &src, |account_data| {
-            account_data.decrease_liquidity_by(validated_transfers.total_amount());
+            account_data.decrease_amount_by(validated_transfers.total_amount());
         });
     }
 
     pub(crate) fn current_block() -> T::BlockNumber {
         <frame_system::Module<T>>::block_number()
     }
 
+    pub(crate) fn ensure_can_init_sale(
+        token_id: T::TokenId,
+        token_data: &TokenDataOf<T>,
+        sale_params: &TokenSaleParamsOf<T>,
+    ) -> DispatchResult {
+        let current_block = Self::current_block();
+
+        // Ensure token offering state is Idle
+        OfferingStateOf::<T>::ensure_idle_of::<T>(token_data)?;
+
+        // Ensure no unrecovered tokens remaining from the previous sale
+        ensure!(
+            token_data.last_sale_remaining_tokens().is_zero(),
+            Error::<T>::RemainingUnrecoveredTokensFromPreviousSale
+        );
+
```

Comment:
> should we allow cap_per_member:is_some() even if token is not permissioned? It does not make sense in that case, and it can help app makers if you have some loud error handling for those sorts of cases.

I thought `cap_per_member` should be independent from the token permission model. My reasoning as to why it makes some sense was that membership has a cost and potentially there could be some additional screening implemented for members, so it's not that trivial to just spawn an infinite amount of identities (as is the case with accounts), therefore there is **_some_** utility in having this cap, but probably not much unless there is some really solid protection agains sybil memberships.

Reply:
`MinSaleDuration` introduced in https://github.com/Joystream/joystream/pull/3618/commits/d8e70c04967c6c8050a12b204be1f39974e955ef, please confirm whether I should disallow setting `cap_per_member` for Permissionless tokens.

Reply:
> therefore there is some utility in having this cap, but probably not much unless there is some really solid protection agains sybil memberships

Hmm ok, makes sense.

> MinSaleDuration introduced in

Great.